### PR TITLE
Connection class measurement

### PR DESCRIPTION
**Pull request by aitorvs**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205631232587242/f

### Description
Connection class/latency measurement and reporting.

This PR adds the following NetP classes:
* `ConnectionQuality` - defines the connection quality based on latency (ms)
* `ConnectionQualityStore` - connection latency measurement store
* `ExponentialGeometricAverage` - class that computes a Exponential Moving Average (EMA)
* `ConnectionClassManager` - calculates the connection class by adding latency measurements and applies some hysteresis 
* `VpnLatencySampler` - class that samples latency and reports when connection class is `POOR` or worse

### Steps to test this PR
Test cases [here](https://app.asana.com/0/0/1205663579937311/f)

### CODE REVIEW

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3632** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3632" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3632?utm_source=stack-comment).

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/connectionclass/VpnLatencySampler.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    private suspend fun sampleLatency(): ConnectionQuality? = withContext(dispatcherProvider.io()) {
+        return@withContext networkProtectionRepository.serverDetails?.ipAddress?.let { server ->
+            val latency = latencyMeasurer.measureLatency(server)
+            if (latency >= 0) {
+                // < 0 latency signals an error and we don't report measurement errors
+                connectionClassManager.addLatency(latency.toDouble())
+            } else {
+                // default to 10s latency. Large enough to swing the latency up "fast enough", small enough to recover
+                // once latency is back to normal
+                connectionClassManager.addLatency(10000.0)
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/connectionclass/VpnLatencySampler.kt:80]: Maybe we don't even want to add a fake latency and just report the pixel.

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/pixels/NetworkProtectionPixelNames.kt [hunk 2]
```diff
@@ -47,6 +47,8 @@ enum class NetworkProtectionPixelNames(
     NETP_WG_ERROR_FAILED_TO_LOAD_WG_LIBRARY("m_netp_ev_wireguard_error_unable_to_load_wireguard_library_c", enqueue = true),
     NETP_WG_ERROR_FAILED_TO_LOAD_WG_LIBRARY_DAILY("m_netp_ev_wireguard_error_unable_to_load_wireguard_library_d", enqueue = true),
     NETP_LATENCY_REPORT("m_netp_ev_latency_c"),
+    NETP_REPORT_POOR_LATENCY_DAILY("m_netp_ev_poor_latency_d"),
```
Comment (karlenDimla) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/pixels/NetworkProtectionPixelNames.kt:50]: We should probably add this in https://app.asana.com/0/1203137811378537/1203882129166495/f

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/connectionclass/ConnectionClassManager.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                // If current quality is UNKNOWN, then changing is always valid.
+                return true
+            }
+        }
+
+        val average = latencyMeasurements.average
+        return if (average > topOfBand * HYSTERESIS_TOP_MULTIPLIER) {
+            true
+        } else if (average < bottomOfBand * HYSTERESIS_BOTTOM_MULTIPLIER) {
+            return true
```
Comment (karlenDimla) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/connectionclass/ConnectionClassManager.kt:131]: NIt: Unnecessary return

Comment (aitorvs) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/connectionclass/ConnectionClassManager.kt:131]: nice, will remove

**Pull request merged**